R/datatypes-s3-base.R

```diff
@@ -1,3 +1,65 @@
+#' @export
+`$.polars_dtype` <- function(x, name) {
+  member_names <- ls(x)
+  method_names <- names(polars_datatype__methods)
+
+  if (name %in% member_names) {
+    get(name, envir = x)
+  } else if (name %in% method_names) {
+    fn <- polars_datatype__methods[[name]]
+    self <- x
+    environment(fn) <- environment()
+    fn
+  } else {
+    NextMethod()
+  }
+}
+
+#' @export
+.DollarNames.polars_dtype <- function(x, pattern = "") {
+  member_names <- ls(x)
+  method_names <- names(polars_datatype__methods)
+
+  all_names <- union(member_names, method_names)
+  filtered_names <- all_names[grep(pattern, all_names)]
+
+  filtered_names[!startsWith(filtered_names, "_")]
+}
+
+#' @export
+`$.polars_dtype_enum` <- function(x, name) {
+  # Enum only method `union`
+  if (identical(name, "union")) {
+    fn <- function(other) {
+      wrap({
+        check_polars_dtype(other)
+        if (!inherits(other, "polars_dtype_enum")) {
+          abort("`other` must be a Enum data type")
+        }
+
+        PlRDataType$new_enum(unique(c(self$categories, other$categories)))
+      })
+    }
+    self <- x
+    environment(fn) <- environment()
+    fn
+  } else {
+    NextMethod()
+  }
+}
+
+#' @export
+.DollarNames.polars_dtype_enum <- function(x, pattern = "") {
+  member_names <- ls(x)
+  # Enum only method `union`
+  method_names <- c("union", names(polars_datatype__methods))
+
+  all_names <- union(member_names, method_names)
+  filtered_names <- all_names[grep(pattern, all_names)]
+
+  filtered_names[!startsWith(filtered_names, "_")]
+}
+
```

Comment:
I'm confused about this, why do we have a special case for enums?

Reply:
This is because only Enum has a method called `union`.